### PR TITLE
Refactor to use block hash as block id

### PR DESCRIPTION
**Pull request by fishseabowl**
## PR Checklist

- [x] I've built the code `cargo build`. For major changes, `cargo build --workspace --release` is recommended.
- [x] I've run the automated unit tests `cargo test`. This executes our automated unit tests.
- [x] I've not broken any existing tests or functionality. In addition to `cargo test`, I've run `cargo clippy`
- [x] I've not introduced any new known security vulnerabilities. I've run `cargo audit`

- [x] I've included a brief description and a good title of the proposed changes and how to test them.
- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've associated an [issue](https://github.com/pyrsia/pyrsia/issues) with this Pull Request. 
- [x] I've checked that the code is up-to-date with the `pyrsia/main` branch.
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/good_pr.md)
- [x] I've assigned this Pull Request to "pyrsia/collaborators"

## Description

Fixes pyrsia/pyrsia#407

This PR refactor the code to use block hash as block id

### CODE REVIEW

**Review by prince-chrismc — COMMENTED** [review]
Please delete the `.db` files

File: src/blockchain/src/header.rs [hunk 1]
```diff
@@ -26,15 +26,21 @@ pub struct HashDigest {
 pub type Address = HashDigest;
 
 // struct Header define the header of a block
+//      parent_hash: 256bit Keccak Hash of the Parent Block(previous Block idx)
+//      committer:  the committer node's PeerID
+//      transactions_root: 256bit Keccak Hash of the root node of Transaction Tries
+//      timestamp: Unix tim, the number of seconds that have elapsed since the Unix epoch, excluding leap seconds
+//      number: block sequence number, the current block number should be the parent(previous) block number plus 1
+//      idx: 256bit Keccak Hash of the Current Block Header, excluding itself
```
Comment (prince-chrismc) [inline src/blockchain/src/header.rs:34]: what does `idx` stand for?

File: src/blockchain/src/header.rs [hunk 2]
```diff
@@ -26,15 +26,21 @@ pub struct HashDigest {
...
-    pub committer: Address,      //the committer node's PeerID
-    pub transactions_root: HashDigest, //256bit Keccak Hash of the root node of Transaction Tries
+    pub parent_hash: HashDigest,
+    pub committer: Address,
+    pub transactions_root: HashDigest,
     pub timestamp: u64,
     pub number: u128,
     pub nonce: u128,
-    pub current_hash: HashDigest, //256bit Keccak Hash of the Current Block Header, excluding itself
+    pub idx: HashDigest,
```
Comment (prince-chrismc) [inline src/blockchain/src/header.rs:43]:
```suggestion
    pub hash: HashDigest,
```

File: src/blockchain/src/header.rs [hunk 3]
```diff
@@ -26,15 +26,21 @@ pub struct HashDigest {
 pub type Address = HashDigest;
 
 // struct Header define the header of a block
+//      parent_hash: 256bit Keccak Hash of the Parent Block(previous Block idx)
```
Comment (prince-chrismc) [inline src/blockchain/src/header.rs:29]: Great improvement, but keep going!

these should be inlined and made as rustdoc 👍 

https://doc.rust-lang.org/rust-by-example/meta/doc.html#inline

File: src/blockchain/src/header.rs [hunk 4]
```diff
@@ -26,15 +26,21 @@ pub struct HashDigest {
...
+//      committer:  the committer node's PeerID
+//      transactions_root: 256bit Keccak Hash of the root node of Transaction Tries
```
Comment (prince-chrismc) [inline src/blockchain/src/header.rs:31]: What is "Transaction Tries" 🤔

**Comment by fishseabowl** [comment]
> Please delete the `.db` files

ok. It's better to add it to .gitignore.

File: src/blockchain/src/header.rs (hunk 1 above)
Comment (fishseabowl) [reply src/blockchain/src/header.rs:34]: Just follow issue #407, idx is block id

File: src/blockchain/src/header.rs (hunk 3 above)
Comment (fishseabowl) [reply src/blockchain/src/header.rs:29]: ok. Thanks

File: src/blockchain/src/header.rs (hunk 4 above)
Comment (fishseabowl) [reply src/blockchain/src/header.rs:31]: We support many transactions in a block, but at the moment we only implement one transaction in a block.  Transaction Tries is Transactions Merkle tree.

File: src/blockchain/src/header.rs (hunk 1 above)
Comment (prince-chrismc) [reply src/blockchain/src/header.rs:34]: please rename the extra "x" is confusing

It could just be called `hash`

**Comment by prince-chrismc** [comment]
> > Please delete the `.db` files
> 
> ok. It's better to add it to .gitignore.

No we have other issues like #82 that are duplicated. That just hides the problem

File: src/blockchain/src/header.rs [hunk 5]
```diff
@@ -22,16 +22,22 @@ use super::crypto::hash_algorithm::HashDigest;
 
 pub type Address = HashDigest;
 
-// struct Header define the header of a block
+/// struct Header define the header of a block
+///      parent_hash: 256bit Keccak Hash of the Parent Block(previous Block hash)
```
Comment (prince-chrismc) [inline src/blockchain/src/header.rs:26]: This is not the correct format for rustdoc, these need to be moved above the fields in question

Example

```rust
pub struct Header {
    /// 256-bit Keccak Hash of the Parent Block (previous Block hash)
    pub parent_hash: HashDigest,
```

Comment (fishseabowl) [reply src/blockchain/src/header.rs:26]: Thanks! 🙏

**Review by prince-chrismc — APPROVED** [review]

File: src/blockchain/src/header.rs [hunk 6]
```diff
@@ -22,16 +22,22 @@ use super::crypto::hash_algorithm::HashDigest;
...
 #[derive(Serialize, Deserialize, Debug, Clone, Hash, PartialEq, Eq, Copy)]
 pub struct Header {
-    pub parent_hash: HashDigest, //256bit Keccak Hash of the Parent Block
-    pub committer: Address,      //the committer node's PeerID
-    pub transactions_root: HashDigest, //256bit Keccak Hash of the root node of Transaction Tries
+    ///      parent_hash: 256bit Keccak Hash of the Parent Block(previous Block hash)
+    pub parent_hash: HashDigest,
+    ///      committer:  the committer node's PeerID
+    pub committer: Address,
+    ///      transactions_root: 256bit Keccak Hash of the root node of Transactions Merkle tree
```
Comment (chb0github) [inline src/blockchain/src/header.rs:32]: Thank you for documenting these - you only need 2 `//`

**Review by chb0github — APPROVED** [review]

**Pull request merged**